File: hooks/_repo.py
"""Shared repository-root resolution for OMCS hooks.

Hooks may run from the source checkout, from a copied local plugin, or with a
workspace environment override. The repo/plugin root is the directory that owns
the checked-in OMCS payload, not the parent of `hooks/`.
"""
from __future__ import annotations

import os
from pathlib import Path


def _is_repo_root(path: Path) -> bool:
    return (
        (path / "hooks" / "hooks.json").is_file()
        and (path / "agents").is_dir()
        and (path / "skills").is_dir()
        and (path / ".cursor-plugin" / "plugin.json").is_file()
    )


def resolve_repo_root(anchor: str | Path) -> Path:
    """Return the OMCS payload root for a hook module."""
    for raw in (
        os.environ.get("OH_MY_CURSOR_WORKSPACE", ""),
        os.getcwd(),
    ):
        if not raw:
            continue
        candidate = Path(raw).expanduser().resolve()
        if _is_repo_root(candidate):
            return candidate

    current = Path(anchor).resolve()
    for candidate in (current.parent, *current.parents):
        if _is_repo_root(candidate):
            return candidate
    return current.parents[1]


def resolve_workspace_root(anchor: str | Path) -> Path:
    """Return the active workspace root used for runtime state files."""
    for raw in (
        os.environ.get("OH_MY_CURSOR_WORKSPACE", ""),
        os.getcwd(),
    ):
        if not raw:
            continue
        candidate = Path(raw).expanduser().resolve()
        if "plugins/local/oh-my-cursor" not in candidate.as_posix():
            return candidate
    return resolve_repo_root(anchor)

File: hooks/_repo.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';

function isRepoRoot(dir: string): boolean {
  try {
    const hooksJson = path.join(dir, 'hooks', 'hooks.json');
    const agentsDir = path.join(dir, 'agents');
    const skillsDir = path.join(dir, 'skills');
    const pluginJson = path.join(dir, '.cursor-plugin', 'plugin.json');

    return (
      fs.existsSync(hooksJson) &&
      fs.statSync(hooksJson).isFile() &&
      fs.existsSync(agentsDir) &&
      fs.statSync(agentsDir).isDirectory() &&
      fs.existsSync(skillsDir) &&
      fs.statSync(skillsDir).isDirectory() &&
      fs.existsSync(pluginJson) &&
      fs.statSync(pluginJson).isFile()
    );
  } catch {
    return false;
  }
}

export function resolveRepoRoot(anchor: string): string {
  const envWorkspace = process.env.OH_MY_CURSOR_WORKSPACE || '';
  const pwd = process.cwd();

  for (const raw of [envWorkspace, pwd]) {
    if (!raw) continue;
    const candidate = path.resolve(raw);
    if (isRepoRoot(candidate)) {
      return candidate;
    }
  }

  let current = path.resolve(anchor);
  try {
    if (fs.statSync(current).isFile()) {
      current = path.dirname(current);
    }
  } catch {
    // ignore
  }

  let candidate = current;
  while (true) {
    if (isRepoRoot(candidate)) {
      return candidate;
    }
    const parent = path.dirname(candidate);
    if (parent === candidate) {
      break;
    }
    candidate = parent;
  }

  return path.resolve(current, '..', '..');
}

export function resolveWorkspaceRoot(anchor: string): string {
  const envWorkspace = process.env.OH_MY_CURSOR_WORKSPACE || '';
  const pwd = process.cwd();

  for (const raw of [envWorkspace, pwd]) {
    if (!raw) continue;
    const candidate = path.resolve(raw);
    if (!candidate.includes('plugins/local/oh-my-cursor')) {
      return candidate;
    }
  }
  return resolveRepoRoot(anchor);
}
